### ISSUE TITLE
Enumerating actions without state

### ISSUE THREAD

**Issue by MichaelGreisberger** [jonathan-laurent/AlphaZero.jl#38]
Probably I am misunderstanding the concept of how to implement a game for AlphaZero.jl but I am having a hard time to understand, how i can enumerate possible actions with only the static data of GameSpec. 
From what I understand, the function `actions(::AbstractGameSpec)` should return all possible actions of the game, while `actions_mask(::AbstractGameEnv)`returns a boolean mask indicating which of those actions are legal in the current state. 
A game with a very limited amount of possible actions is fairly easy to implement in this setup. But how would I, for example, implement Chess? For Chess the state determines which actions are possible and as there are many possible actions it seams inefficient to enumerate them all and pass a boolean mask indicating which are playable. But than again I am probably misunderstanding something. 

Hope someone can get me on track. Would really love to use this package!

**Comment by jonathan-laurent** [comment]
Thanks for your interest in AlphaZero.jl.
I think you understood the game interface correctly but you are overestimating how inefficient it is.

First of all, you do not have to enumerate all available actions and check their availability every time `actions_mask` is called. You can cache this mask in your `GameEnv` object and only update a few components every time an action is played. Also, you can use a sparse boolean vector instead of a dense one.

That being said, I am not even sure that these kind of optimizations would buy you that much anyway. Indeed, game simulation speed is everything when writing a minmax or alpha-beta agent (e.g. Stockfish); however, with AlphaZero, the time needed to simulate the game or compute available actions is vastly dominated by the time spent in neural network inference anyway. (In contrast, I think it is much more important to reduce the memory footprint of storing a state as running many instances of MCTS in parallel can consume a lot of memory).

Right now, the main reason for requiring the user to provide an `actions_mask` function (in contrast to just having an `is_valid` function) is that it enables masking out invalid actions returned by the neural network. Also, it makes it easy to represent the resulting policy as a probability vector of size `length(available_actions(state))`.

I guess we could avoid doing this and just allow the neural network to suggest invalid actions and just reject them as they are attempted. But I fear it would significantly complicate the implementation for little to no gain.

That being said, I am really open to a debate on this and I would welcome new ideas. More generally, I believe that the current game interface could be generalized to allow for more complex and structured action spaces. For example, imagine a game where an action is represented as a tuple of twenty boolean values and most actions are typically available (so we cannot just use sparse vectors). A neural network could still output a policy on this action space (e.g. by producing an independent probability for each component or by using an RNN with beam search to generate a collection of candidate actions with associated scores) but the `actions_mask` mechanism would not scale here.

**Comment by MichaelGreisberger** [comment]
Thanks for your very fast and comprehensive answer. The main concern I have is, that it is hard to enumerate all possible actions of the game without a state. This is because of the rules of the game (Fanorona) which is best compared to checkers where you basically have very simple actions that can be combined to long capturing sequences. The rules of Fanorona are a little more complex than those of checkers and also provide more simple moves. 

In fact what I would need is to implement `available_actions(::AbstractGameEnv)` which I guess currently is a combination of  `actions(::AbstractGameSpec)` and `actions_mask(::AbstractGameEnv)`. This way I could generate all possible actions (including simple and complex ones) from the current state. The resulting list of actions would not be of the same size for every state (although I don't think this is necessary). 

Another possibility would be to only use simple actions, as those can be enumerated easily, and run AlphaZero subsequently several times in the same turn until a turn-ending action is selected (this can either be an action with no possible successive action or pass). What I don't like about this solution is, that it is getting much harder to compare the resulting player to other MCTS/AlphaBeta agents, as those (at least for Fanorona) would select a move as a single action regardless of whether its a simple or a complex one. Of course comparing with those agents would be unfair regardless of those choices, as the MCTS/AlphaBeta agents that i would compare to dont use concurrent rollouts or GPU's.

**Comment by jonathan-laurent** [comment]
I am not sure I understand what your objection is to using simple actions. AlphaZero.jl perfectly accommodates a player playing several times in a row and this is usually the best way to model action sequences. An algorithm like MinMax or AlphaBeta could easily support this too (and our Minmax implementation in fact does).

Also, don't forget that your neural network will have to output a probability distribution over all available actions. The simplest way to do this by far (which is the only way currently supported) is for the network to take a state as an input and output a vector of logits of length `length(actions(game_spec))`, where logits corresponding to invalid actions are then masked out.

What else would you do if you weren't able to enumerate all potential game actions? Would you have a network that takes a state and an action as inputs and then returns a score for it? And then you would evaluate it on all available actions as returned by your function? In this case though, computing the action distribution might be more costly and the network would probably have to be trained a bit differently.

**Comment by MichaelGreisberger** [comment]
Well its not really an objection probably more of a habit, as I implemented my MCTS based agents that way (computing all possible moves from a state and selecting the best of those after some rollouts). The idea was to implement an AlphaZero agent for this game and to evaluate it against my MCTS agents. The MCTS agents use time as a computational boundry (e.g. they have 5 seconds time to think of their next move). I wanted to implement the same restrictions for AlphaZero which gets harder if I don't know the number of moves the agent takes in a single turn. 

In my initial comment I did not know why the agent needed the fixed size array but now it makes perfect sense. My initial idea was close to your last suggestion - input a state and the list of possible actions into the neural network and output a probability distribution over those actions. But I guess it would be difficult to build and train a network like that. So most likely it would be best to stick to the structure supported by AlphaZero.jl. The number of simple actions in Fanorona is about 600 (rough estimate), do you think this is a problem?

**Comment by oscardssmith** [comment]
That's fine. Most chess implementations have 1800 or so.

**Comment by jonathan-laurent** [comment]
As @oscardssmith said, having 600 potential actions should be fine.

The more important question here is: how many actions are typically available in each state? Indeed, games with a high branching factor take much more time to solve using AlphaZero.jl unless you start with an initial policy that already provides a strong prior on what actions are worth exploring. For example, it would not be realistic to use AlphaZero to learn chess from scratch on a single computer (it would be interesting to use supervised learning to learn a reasonable initial policy and then improve it using AlphaZero though).

**Comment by MichaelGreisberger** [comment]
The average branching factor is about 12 but there are situations where the branching factor gets above 100. I could use a big cluster to train the network but I have never worked with it before and I don't know much about it yet. Supervised learning would be possible as I have a set of roughly 100,000 games of different agents.

**Comment by jonathan-laurent** [comment]
@MichaelGreisberger This sounds exciting. I hope AlphaZero.jl is useful to you and please don't hesitate to share updates!
